services/catalog/src/simcore_service_catalog/core/events.py

```diff
@@ -1,65 +1,90 @@
+from fastapi_lifespan_manager import LifespanManager, State
+from servicelib.fastapi.postgres_lifespan import (
+    PostgresLifespanStateKeys,
+    postgres_lifespan_manager,
+)
+from servicelib.fastapi.prometheus_instrumentation import (
+    lifespan_prometheus_instrumentation,
+)
+from servicelib.fastapi.tracing import initialize_tracing
+
+from .._meta import APP_FINISHED_BANNER_MSG, APP_NAME, APP_STARTED_BANNER_MSG
+from ..api.rpc.routes import setup_rpc_api_routes
+from ..db.events import setup_database
+from ..services.director import setup_director
+from ..services.function_services import setup_function_services
+from ..services.rabbitmq import setup_rabbitmq
+from .background_tasks import setup_background_task
+from .settings import ApplicationSettings
 
 _logger = logging.getLogger(__name__)
 
 
-EventCallable: TypeAlias = Callable[[], Awaitable[None]]
-
-
-def _flush_started_banner() -> None:
+def flush_started_banner() -> None:
     # WARNING: this function is spied in the tests
     print(APP_STARTED_BANNER_MSG, flush=True)  # noqa: T201
 
 
-def _flush_finished_banner() -> None:
+def flush_finished_banner() -> None:
     print(APP_FINISHED_BANNER_MSG, flush=True)  # noqa: T201
 
 
-def create_on_startup(app: FastAPI) -> EventCallable:
-    async def _() -> None:
-        _flush_started_banner()
+async def _setup_app(app: FastAPI) -> AsyncIterator[State]:
+    flush_started_banner()
+
+    settings: ApplicationSettings = app.state.settings
+
+    if settings.CATALOG_TRACING:
+        initialize_tracing(app, settings.CATALOG_TRACING, APP_NAME)
+
+    yield {
+        PostgresLifespanStateKeys.POSTGRES_SETTINGS: settings.CATALOG_POSTGRES,
+        "prometheus_instrumentation_enabled": settings.CATALOG_PROMETHEUS_INSTRUMENTATION_ENABLED,
+    }
+
+    flush_finished_banner()
+
+
+async def _setup_prometheus_instrumentation_adapter(
+    app: FastAPI, state: State
+) -> AsyncIterator[State]:
+    enabled = state.get("prometheus_instrumentation_enabled", False)
+    if enabled:
+        async for prometheus_state in lifespan_prometheus_instrumentation(app):
+            yield prometheus_state
+
 
-        # setup connection to pg db
-        if app.state.settings.CATALOG_POSTGRES:
-            await connect_to_db(app, app.state.settings.CATALOG_POSTGRES)
-            await setup_default_product(app)
+def create_app_lifespan():
+    # app lifespan
+    app_lifespan = LifespanManager()
+    app_lifespan.add(_setup_app)
 
-        if app.state.settings.CATALOG_DIRECTOR:
-            # setup connection to director
-            await setup_director(app)
+    # WARNING: order matters
 
-            # FIXME: check director service is in place and ready. Hand-shake??
-            # SEE https://github.com/ITISFoundation/osparc-simcore/issues/1728
-            await start_registry_sync_task(app)
+    # - postgres lifespan
+    postgres_lifespan_manager.add(setup_database)
+    app_lifespan.include(postgres_lifespan_manager)
```

Comment:
This is nice, now I get it why you have a function where you define the layout of the lifespans.